File: packages/react-native/test/retained/expo-probe/android/src/main/java/dev/taurinative/expoprobe/RetainedExpoProbeModule.kt
package dev.taurinative.expoprobe

import android.Manifest
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.atomic.AtomicInteger

object ProbeState {
  val created = AtomicInteger(); val destroyed = AtomicInteger(); val callbacks = AtomicInteger()
  val applicationCreates = AtomicInteger(); val activityCreates = AtomicInteger()
  val resumes = AtomicInteger(); val pauses = AtomicInteger(); val intents = AtomicInteger(); val backs = AtomicInteger()
  fun snapshot() = mapOf("created" to ProbeState.created.get(), "destroyed" to ProbeState.destroyed.get(), "callbacks" to ProbeState.callbacks.get(),
        "applicationCreates" to ProbeState.applicationCreates.get(), "activityCreates" to ProbeState.activityCreates.get(),
        "resumes" to ProbeState.resumes.get(), "pauses" to ProbeState.pauses.get(), "intents" to ProbeState.intents.get(), "backs" to ProbeState.backs.get())
}
class RetainedExpoProbeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RetainedExpoProbe")
    OnCreate { ProbeState.created.incrementAndGet() }
    OnDestroy { ProbeState.destroyed.incrementAndGet() }
    Function("snapshot") {
      ProbeState.snapshot()
    }
    AsyncFunction("requestLocation") { promise: Promise ->
      checkNotNull(appContext.permissions).askForPermissions({ result ->
        ProbeState.callbacks.incrementAndGet()
        promise.resolve(result.mapValues { it.value.status.name.lowercase() })
      }, Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
    }
  }
}

File: packages/react-native/test/retained/expo-probe/ios/RetainedExpoProbeModule.swift
import ExpoModulesCore

@objc(TNExpoProbeState)
public final class ProbeState: NSObject, @unchecked Sendable {
  @objc public static let shared = ProbeState()
  private let lock = NSLock()
  private var counts: [String: Int] = [:]
  func increment(_ name: String) { lock.lock(); defer { lock.unlock() }; counts[name, default: 0] += 1 }
  @objc public func snapshot() -> [String: Int] { lock.lock(); defer { lock.unlock() }; return counts }
}

public class RetainedExpoProbeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RetainedExpoProbe")
    OnCreate { ProbeState.shared.increment("created") }
    OnDestroy { ProbeState.shared.increment("destroyed") }
    Function("snapshot") { ProbeState.shared.snapshot() }
    AsyncFunction("appCallbacks") { (promise: Promise) in
      MainActor.assumeIsolated {
      let app = UIApplication.shared
      guard let delegate = app.delegate,
        delegate.responds(to: #selector(UIApplicationDelegate.applicationDidReceiveMemoryWarning(_:))),
        delegate.responds(to: #selector(UIApplicationDelegate.application(_:performFetchWithCompletionHandler:))) else {
        promise.reject("missing_callback", "Original Tauri delegate did not expose Expo callbacks")
        return
      }
      delegate.applicationDidReceiveMemoryWarning?(app)
      delegate.application?(app, performFetchWithCompletionHandler: { result in
        ProbeState.shared.increment("fetchCompletions")
        promise.resolve(["result": result.rawValue, "counts": ProbeState.shared.snapshot()])
      })
      }
    }.runOnQueue(.main)
  }
}

public class RetainedExpoProbeSubscriber: ExpoAppDelegateSubscriber {
  public func appDelegateWillBeginInitialization() { ProbeState.shared.increment("applicationCreates") }
  public func application(_ app: UIApplication, willFinishLaunchingWithOptions options: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    ProbeState.shared.increment("willLaunch"); return true
  }
  public func application(_ app: UIApplication, didFinishLaunchingWithOptions options: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    ProbeState.shared.increment("didLaunch"); return true
  }
  public func applicationDidBecomeActive(_ app: UIApplication) { ProbeState.shared.increment("resumes") }
  public func applicationWillResignActive(_ app: UIApplication) { ProbeState.shared.increment("pauses") }
  public func applicationDidEnterBackground(_ app: UIApplication) { ProbeState.shared.increment("backgrounds") }
  public func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
    ProbeState.shared.increment("urls"); return false
  }
  public func application(_ app: UIApplication, continue activity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool {
    ProbeState.shared.increment("activities"); return false
  }
  public func customizeRootView(_ rootView: UIView) { ProbeState.shared.increment("customizations") }
  public func applicationDidReceiveMemoryWarning(_ app: UIApplication) { ProbeState.shared.increment("memoryWarnings") }
  public func application(_ app: UIApplication, performFetchWithCompletionHandler completion: @escaping (UIBackgroundFetchResult) -> Void) {
    ProbeState.shared.increment("fetches"); completion(.newData)
  }
}
